pkg/util/ipvs/kernelcheck_linux.go

```diff
@@ -63,29 +65,46 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
 	loadModules.Insert(mods...)
 	modules := wantModules.Difference(loadModules).UnsortedList()
 
-	// Check builtin modules exist or not
-	if len(modules) != 0 {
-		builtinModsFilePath := fmt.Sprintf("/lib/modules/%s/modules.builtin", kernelVersion)
-		out, err := r.Executor.Command("cut", "-f1", "-d", " ", builtinModsFilePath).CombinedOutput()
+	if len(modules) == 0 {
+		return nil, nil
+	}
+
+	// Check if loadable modules are installed
+	var missingModules []string
+	for _, module := range modules {
+		out, err = r.Executor.Command("modinfo", module).CombinedOutput()
 		if err != nil {
-			errors = append(errors, fmt.Errorf("error getting required builtin kernel modules: %v(%s)", err, out))
-			return nil, errors
+			warnings = append(warnings, pkgerrors.Wrapf(err, "error getting module info for %q; output: %s", module, strings.TrimSpace(string(out))))
+			missingModules = append(missingModules, module)
 		}
+	}
 
-		builtInModules := sets.NewString()
-		for _, builtInMode := range ipvsModules {
-			match, _ := regexp.Match(builtInMode+".ko", out)
-			if !match {
-				builtInModules.Insert(string(builtInMode))
-			}
-		}
-		if len(builtInModules) != 0 {
-			warnings = append(warnings, fmt.Errorf(
-				"the IPVS proxier will not be used, because the following required kernel modules are not loaded: %v or no builtin kernel ipvs support: %v\n"+
-					"you can solve this problem with following methods:\n 1. Run 'modprobe -- ' to load missing kernel modules;\n"+
-					"2. Provide the missing builtin kernel ipvs support\n", modules, builtInModules))
+	if len(missingModules) == 0 {
+		return warnings, nil
+	}
+
+	// Check if builtin modules exist
+	builtinModsFilePath := fmt.Sprintf("/lib/modules/%s/modules.builtin", kernelVersion)
+	out, err = r.Executor.Command("cut", "-f1", "-d", " ", builtinModsFilePath).CombinedOutput()
```

Comment:
Can we get rid of this by replacing it with simple read+parse function?

Reply:
I'd be happy to do that if we reach overall agreement to accept this patch with @Lion-Wei and @stewart-yu

Reply:
@neolit123 

> please close this PR as it seems the PR replaces it.

Let's give @Lion-Wei and @stewart-yu another chance to accept it. I still think it's a correct fix and after @kad's comment I don't see any reason of not accepting it.

Reply:
+1 would be great if we can use some of the existing util functions for checking these i.e. https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L129-L138